ledger/complete/mtrie/trie/trie.go

```diff
@@ -324,13 +363,39 @@ func update(
 				// the case where the recursion stops: only one path to update
 				if len(paths) == 1 {
 					if !parentNode.Payload().Equals(&payloads[i]) {
-						return node.NewLeaf(paths[i], payloads[i].DeepCopy(), nodeHeight)
+						n = node.NewLeaf(paths[i], payloads[i].DeepCopy(), nodeHeight)
+
+						allocatedRegCountDelta = 0
+						if payloads[i].IsEmpty() {
+							// Old payload is not empty while new payload is empty.
+							// Allocated register will be unallocated.
+							allocatedRegCountDelta = -1
```

Comment:
I am not sure this logic here is correct in all cases. Depending on whether the `prune` parameter is set or not, we might add registers with empty `Value` 

Consider the following sequence of events:
1. a register with empty Value is added with a `NewTrieWithUpdatedRegisters` call with `prune = false`
  https://github.com/onflow/flow-go/blob/ae98303292d312358fbe9b17d0c0abab4babc51a/ledger/complete/mtrie/trie/trie.go#L350-L353
  this will _not_ increase the `regCount`
3. In a second call, with `NewTrieWithUpdatedRegisters` call with `prune = true` we remove the leaf. This will decrease `regCount`.

In my view, the overall complexity here originates from the `prune` flag. You can't make any assumption whether the leaf represents an allocated or an unallocated register, because with `prune = false`, our trie will store unallocasted registers.  Hence, whenever you change the value (meaning the register was already explicitly represented in the Trie _before_ the update), you have to consider all 4 possibilities:
`{Before the update, the leaf represented an allocated or unallocated register?}` x `{After the update, the leaf represents an allocated or unallocated register?}`

At the moment you only count the allocated registers. Thereby, whenever dealing with leafs that might potentially represent unallocated registers, you need extra logic. 
This entire consideration would be much simplified, if you changed the convention: just count _all_ registers that are represented by leafs in the trie (no matter whether they are allocated or unallocated). This would be [Option 2](https://github.com/onflow/flow-go/pull/2126#pullrequestreview-909500380). This will considerably simplify the implementation and the extra complexity we add to an already complex algorithm.

Reply:
If you desire to really only count allocated registers, you could encapsulate the respective logic to compute the deltas in a separate function, which would help the clarity of the code
```golang
func computeDeltas(oldNode, newNode *node.Node) (allocatedRegCountDelta int64, allocatedRegSizeDelta int64) {
	var oldNodeIsAllocated, newNodeIsAlocated int64
	var oldNodeSize, newNodeSize int
	if oldNode != nil && !oldNode.Payload().IsEmpty() {
		oldNodeIsAllocated = 1
		oldNodeSize = oldNode.Payload().Size()
	}
	if newNode != nil && !newNode.Payload().IsEmpty() {
		newNodeIsAlocated = 1
		newNodeSize = newNode.Payload().Size()
	}

	allocatedRegCountDelta = newNodeIsAlocated - oldNodeIsAllocated
	allocatedRegSizeDelta = int64(newNodeSize - oldNodeSize)
	return
}

```

Reply:
> I am not sure this logic here is correct in all cases. Depending on whether the prune parameter is set or not, we might add registers with empty Value
>
> Consider the following sequence of events:
>
> 1.  a register with empty Value is added with a NewTrieWithUpdatedRegisters call with prune = false
>...
>     this will not increase the regCount
> 2. In a second call, with NewTrieWithUpdatedRegisters call with prune = true we remove the leaf. This will decrease regCount.

I think the logic correctly handles the sequence of events mentioned.  Because in step 2, we don't decrease the reg count when `prune = true`.

To confirm, I added `TestTrieAllocatedRegCountRegSizeWithMixedPruneFlag` just now in d9573d0cb040ba85c3cef19ab5009e2112614686 and cbb8805ac45cfb8b4c5fd8b6d76f964f15f3f598.

```
TestTrieAllocatedRegCountRegSizeWithMixedPruneFlag tests allocated
register count and size for updated trie with mixed pruning flag.
It tests the following updates:

* step 1 : update empty trie with new paths and payloads
           (255 allocated registers)

* step 2 : remove a payload without pruning
           (254 allocated registers)

* step 3a: remove previously removed payload with pruning
           (254 allocated registers)

* step 3b: update trie from step 2 with a new payload (sibling of
           removed payload) with pruning (255 allocated registers)
```

Reply:
>You could extract the logic into a sub-function, which would also increase readability of the code:
>
> `func computeDeltas(...`

I agree and extracted to two functions in 4bcc1c815d300bdee1af5b5429f7b19fb4394928

```Go
// allocatedRegDeltasFromHigherHeight returns the deltas needed
// to compute the allocated reg count and reg size when a payload
// needs to be updated or unallocated at a lower height.
func allocatedRegDeltasFromHigherHeight(oldPayload *ledger.Payload) (allocatedRegCountDelta, allocatedRegSizeDelta int64)

// allocatedRegDeltas returns the allocated reg count and reg size deltas
// computed from old payload and new payload.
// CAUTION: !oldPayload.Equals(newPayload)
func allocatedRegDeltas(oldPayload, newPayload *ledger.Payload) (allocatedRegCountDelta, allocatedRegSizeDelta int64) 
```